DAOS-5535 control: route storage scan via drpc

tanabarr opened the pull request:
When I/O Server processes are running they claim exclusive access to
an NVMe SSD through SPDK (we don't allow shmem between procs to
enforce isolation). This makes the device unavailable via go-spdk
bindings in daos_admin invocations and device details are therefore
missing from 'dmg storage scan' results. In this situation, send scan
requests via dRPC to all daos_io_server processes to retrieve details
of any claimed devices.

- scan now shows all controllers regardless of I/O process run state
- scan --nvme-health reports controller stats regardless of run state
- scan --nvme-meta reports any SSD-target/rank mappings

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/1/execution/node/245/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/1/execution/node/248/log

daosbuild1 commented:
Test stage Build on CentOS 7 debug completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/1/execution/node/303/log

daosbuild1 commented:
Test stage Build on CentOS 7 release completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/1/execution/node/291/log

daosbuild1 commented:
Test stage Build on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/1/execution/node/258/log

daosbuild1 commented:
Test stage Build on Ubuntu 20.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/1/execution/node/276/log

daosbuild1 commented:
Test stage Build on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/1/execution/node/306/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/1/execution/node/235/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

kjacque commented on src/control/server/ctl_storage_rpc_test.go:
```diff
@@ -213,9 +297,38 @@ func TestServer_CtlSvc_StorageScan(t *testing.T) {
+				}
+
 				cs := mockControlService(t, log, config, tc.bmbc, tc.smbc, nil)
+				if configIdx == 0 {
```
These configIdx-based checks are a little confusing to me. Is there a reason we wouldn't want to use the mock all the time? Or care only about req.ConfigDevicesOnly regardless of which config we're looking at?

tanabarr replied:
tidied this up a little, now passing in Health/Meta (the interesting portion) in that request , so it's a bit more explicit. if it's an empty config then we don't want to override the harness with active instances (the instance list should be empty)

tanabarr replied:
refactored further to remove configIdx-based checks

kjacque reviewed (dismissed):
In general it looks good to me. A couple minor comments - could be addressed in a follow-on if changes are needed.

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3423/3/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3423/3/display/redirect

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/5/execution/node/503/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/6/execution/node/502/log

tanabarr commented on src/control/server/ctl_storage_rpc.go:
```diff
@@ -127,75 +129,185 @@ func (c *StorageControlService) StoragePrepare(ctx context.Context, req *ctlpb.S
+	if empty != "" {
+		return
+	}
+	ms = m + s
```
done

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/8/execution/node/503/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/9/execution/node/502/log

mjmac reviewed (commented):
I started a review but wanted to actually test the PR -- haven't had time to do that yet, but I'll leave some initial feedback based on just reading the code.

tanabarr commented on src/control/cmd/dmg/storage.go:
```diff
@@ -143,7 +148,10 @@ func (cmd *storageScanCmd) Execute(_ []string) error {
 			cmd.log.Debug("--verbose flag ignored if --json specified")
 		}
 		if cmd.NvmeHealth {
-			cmd.log.Debug("--health flag ignored if --json specified")
+			cmd.log.Debug("--nvme-health flag ignored if --json specified")
+		}
+		if cmd.NvmeMeta {
+			cmd.log.Debug("--nvme-meta flag ignored if --json specified")
 		}
```
they are not required . thanks for the catch

tanabarr commented on src/control/cmd/dmg/storage.go:
```diff
@@ -168,6 +176,17 @@ func (cmd *storageScanCmd) Execute(_ []string) error {
 		return resp.Errors()
 	}
+	if cmd.NvmeMeta {
+		if cmd.Verbose {
+			cmd.log.Debug("--verbose flag ignored if --nvme-meta specified")
+		}
```
changed log level to info

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

tanabarr commented on src/control/server/ctl_storage_rpc.go:
```diff
@@ -127,75 +129,185 @@ func (c *StorageControlService) StoragePrepare(ctx context.Context, req *ctlpb.S
+// checkModelSerial verifies we have non-null model and serial identifiers.
+// Returns the concatenated identifier along with name of first empty field
+// and boolean indicating whether both are populated or not.
+func checkModelSerial(m string, s string) (ms, empty string, ok bool) {
```
done

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

tanabarr commented on src/control/cmd/dmg/pretty/storage.go:
```diff
@@ -64,7 +64,47 @@ func PrintNvmeHealthMap(hsm control.HostStorageMap, out io.Writer, opts ...contr
-func printSmdDevice(dev *control.SmdDevice, out io.Writer, opts ...control.PrintConfigOption) error {
+// PrintNvmeMetaMap generates a human-readable representation of the supplied
+// HostStorageMap, with a focus on presenting the NVMe Device Server Meta Data.
+func PrintNvmeMetaMap(hsm control.HostStorageMap, out io.Writer, opts ...control.PrintConfigOption) error {
```
there is now

daosbuild1 commented on src/control/cmd/dmg/pretty/storage_test.go:
```diff
@@ -172,6 +172,176 @@ PCI:%s Model:%s FW:%s Socket:%d Capacity:%s
+		controllerA = storage.MockNvmeController(1)
+		controllerB = storage.MockNvmeController(2)
+		controllerC = storage.MockNvmeController(1)
+		controllerD = storage.MockNvmeController(2)
```
```suggestion
		controlled = storage.MockNvmeController(2)
```

daosbuild1 replied:
```suggestion
		controlled = storage.MockNvmeController(2)
```

daosbuild1 replied:
```suggestion
		controlled = storage.MockNvmeController(2)
```

daosbuild1 replied:
```suggestion
		controlled = storage.MockNvmeController(2)
```

daosbuild1 replied:
```suggestion
		controlled = storage.MockNvmeController(2)
```

daosbuild1 replied:
```suggestion
		controlled = storage.MockNvmeController(2)
```

daosbuild1 replied:
```suggestion
		controlled = storage.MockNvmeController(2)
```

daosbuild1 replied:
```suggestion
		controlled = storage.MockNvmeController(2)
```

daosbuild1 commented on src/control/cmd/dmg/pretty/storage_test.go:
```diff
@@ -172,6 +172,176 @@ PCI:%s Model:%s FW:%s Socket:%d Capacity:%s
+					&control.HostStorage{
+						NvmeDevices: storage.NvmeControllers{
+							controllerC,
+							controllerD,
```
```suggestion
							controlled,
```

daosbuild1 replied:
```suggestion
							controlled,
```

daosbuild1 replied:
```suggestion
							controlled,
```

daosbuild1 replied:
```suggestion
							controlled,
```

daosbuild1 replied:
```suggestion
							controlled,
```

daosbuild1 replied:
```suggestion
							controlled,
```

daosbuild1 replied:
```suggestion
							controlled,
```

daosbuild1 commented on src/control/cmd/dmg/pretty/storage_test.go:
```diff
@@ -172,6 +172,176 @@ PCI:%s Model:%s FW:%s Socket:%d Capacity:%s
+				controllerC.SmdDevices[0].UUID, controllerC.SmdDevices[0].TargetIDs,
+				controllerC.SmdDevices[0].Rank, controllerC.SmdDevices[0].State,
+
+				controllerD.PciAddr, controllerD.Model, controllerD.FwRev,
+				controllerD.SocketID, humanize.Bytes(controllerD.Capacity()),
+				controllerD.SmdDevices[0].UUID, controllerD.SmdDevices[0].TargetIDs,
+				controllerD.SmdDevices[0].Rank, controllerD.SmdDevices[0].State),
```
```suggestion
				controlled.PciAddr, controlled.Model, controlled.FwRev,
				controlled.SocketID, humanize.Bytes(controlled.Capacity()),
				controlled.SmdDevices[0].UUID, controlled.SmdDevices[0].TargetIDs,
				controlled.SmdDevices[0].Rank, controlled.SmdDevices[0].State),
```

daosbuild1 replied:
```suggestion
				controlled.PciAddr, controlled.Model, controlled.FwRev,
				controlled.SocketID, humanize.Bytes(controlled.Capacity()),
				controlled.SmdDevices[0].UUID, controlled.SmdDevices[0].TargetIDs,
				controlled.SmdDevices[0].Rank, controlled.SmdDevices[0].State),
```

daosbuild1 replied:
```suggestion
				controlled.PciAddr, controlled.Model, controlled.FwRev,
				controlled.SocketID, humanize.Bytes(controlled.Capacity()),
				controlled.SmdDevices[0].UUID, controlled.SmdDevices[0].TargetIDs,
				controlled.SmdDevices[0].Rank, controlled.SmdDevices[0].State),
```

daosbuild1 replied:
```suggestion
				controlled.PciAddr, controlled.Model, controlled.FwRev,
				controlled.SocketID, humanize.Bytes(controlled.Capacity()),
				controlled.SmdDevices[0].UUID, controlled.SmdDevices[0].TargetIDs,
				controlled.SmdDevices[0].Rank, controlled.SmdDevices[0].State),
```

daosbuild1 replied:
```suggestion
				controlled.PciAddr, controlled.Model, controlled.FwRev,
				controlled.SocketID, humanize.Bytes(controlled.Capacity()),
				controlled.SmdDevices[0].UUID, controlled.SmdDevices[0].TargetIDs,
				controlled.SmdDevices[0].Rank, controlled.SmdDevices[0].State),
```

daosbuild1 replied:
```suggestion
				controlled.PciAddr, controlled.Model, controlled.FwRev,
				controlled.SocketID, humanize.Bytes(controlled.Capacity()),
				controlled.SmdDevices[0].UUID, controlled.SmdDevices[0].TargetIDs,
				controlled.SmdDevices[0].Rank, controlled.SmdDevices[0].State),
```

daosbuild1 replied:
```suggestion
				controlled.PciAddr, controlled.Model, controlled.FwRev,
				controlled.SocketID, humanize.Bytes(controlled.Capacity()),
				controlled.SmdDevices[0].UUID, controlled.SmdDevices[0].TargetIDs,
				controlled.SmdDevices[0].Rank, controlled.SmdDevices[0].State),
```

daosbuild1 replied:
```suggestion
				controlled.PciAddr, controlled.Model, controlled.FwRev,
				controlled.SocketID, humanize.Bytes(controlled.Capacity()),
				controlled.SmdDevices[0].UUID, controlled.SmdDevices[0].TargetIDs,
				controlled.SmdDevices[0].Rank, controlled.SmdDevices[0].State),
```

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/12/execution/node/61/log

tanabarr commented:
gatekeeper please use PR description and title for the commit message when landing

tanabarr commented:
`dmg storage scan [--verbose|--nvme-meta|--nvme-health]` command output when I/O are running and have claimed NVMe SSDs through SPDK:
```
bash-4.2$ dmg storage scan
Hosts        SCM Total             NVMe Total
-----        ---------             ----------
wolf-[71-72] 6.4 TB (2 namespaces) 3.1 TB (3 controllers)
bash-4.2$ dmg storage scan --verbose
------------
wolf-[71-72]
------------
SCM Namespace Socket ID Capacity
------------- --------- --------
pmem0         0         3.2 TB
pmem1         0         3.2 TB

NVMe PCI     Model                FW Revision Socket ID Capacity
--------     -----                ----------- --------- --------
0000:81:00.0 INTEL SSDPED1K750GA  E2010325    1         750 GB
0000:87:00.0 INTEL SSDPEDMD016T4  8DV10171    1         1.6 TB
0000:da:00.0 INTEL SSDPED1K750GA  E2010325    1         750 GB

bash-4.2$ dmg system query
Rank  State
----  -----
[0-3] Joined

bash-4.2$ dmg storage scan --nvme-meta
-------
wolf-71
-------
PCI:0000:81:00.0 Model:INTEL SSDPED1K750GA  FW:E2010325 Socket:1 Capacity:750 GB
  SMD Devices
    UUID:b1b346fc-fb59-44e4-ae62-ddb57540f14d Targets:[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15] Rank:0 State:NORMAL

PCI:0000:da:00.0 Model:INTEL SSDPED1K750GA  FW:E2010325 Socket:1 Capacity:750 GB
  SMD Devices
    UUID:9233fd96-4715-4268-b6d3-83b371ed94f5 Targets:[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15] Rank:3 State:NORMAL

-------
wolf-72
-------
PCI:0000:81:00.0 Model:INTEL SSDPED1K750GA  FW:E2010435 Socket:1 Capacity:750 GB
  SMD Devices
    UUID:c3c0b375-e06d-4c28-b59c-67cb204b2e60 Targets:[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15] Rank:2 State:NORMAL

PCI:0000:da:00.0 Model:INTEL SSDPED1K750GA  FW:E2010435 Socket:1 Capacity:750 GB
  SMD Devices
    UUID:ffd2a5b7-f744-445e-a020-43bbe2184172 Targets:[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15] Rank:1 State:NORMAL

bash-4.2$ dmg storage scan --nvme-health
-------
wolf-71
-------
PCI:0000:81:00.0 Model:INTEL SSDPED1K750GA  FW:E2010325 Socket:1 Capacity:750 GB
  Health Stats:
    Timestamp:77828-05-15 17:23:28 +0000 UTC
    Temperature:318K(44.85C)
    Controller Busy Time:0s
    Power Cycles:15
    Power On Duration:10402h0m0s
    Unsafe Shutdowns:13
    Error Count:0
    Media Errors:0
    Read Errors:0
    Write Errors:0
    Unmap Errors:0
    Checksum Errors:0
    Error Log Entries:0
  Critical Warnings:
    Temperature: OK
    Available Spare: OK
    Device Reliability: OK
    Read Only: OK
    Volatile Memory Backup: OK

PCI:0000:da:00.0 Model:INTEL SSDPED1K750GA  FW:E2010325 Socket:1 Capacity:750 GB
  Health Stats:
    Timestamp:77828-05-14 08:00:39 +0000 UTC
    Temperature:320K(46.85C)
    Controller Busy Time:0s
    Power Cycles:15
    Power On Duration:10402h0m0s
    Unsafe Shutdowns:13
    Error Count:0
    Media Errors:0
    Read Errors:0
    Write Errors:0
    Unmap Errors:0
    Checksum Errors:0
    Error Log Entries:0
  Critical Warnings:
    Temperature: OK
    Available Spare: OK
    Device Reliability: OK
    Read Only: OK
    Volatile Memory Backup: OK

-------
wolf-72
-------
PCI:0000:81:00.0 Model:INTEL SSDPED1K750GA  FW:E2010435 Socket:1 Capacity:750 GB
  Health Stats:
    Timestamp:77828-09-21 15:53:36 +0000 UTC
    Temperature:316K(42.85C)
    Controller Busy Time:8m0s
    Power Cycles:23
    Power On Duration:10399h0m0s
    Unsafe Shutdowns:18
    Error Count:0
    Media Errors:0
    Read Errors:0
    Write Errors:0
    Unmap Errors:0
    Checksum Errors:0
    Error Log Entries:0
  Critical Warnings:
    Temperature: OK
    Available Spare: OK
    Device Reliability: OK
    Read Only: OK
    Volatile Memory Backup: OK

PCI:0000:da:00.0 Model:INTEL SSDPED1K750GA  FW:E2010435 Socket:1 Capacity:750 GB
  Health Stats:
    Timestamp:77828-09-20 07:17:18 +0000 UTC
    Temperature:320K(46.85C)
    Controller Busy Time:1m0s
    Power Cycles:23
    Power On Duration:10399h0m0s
    Unsafe Shutdowns:19
    Error Count:0
    Media Errors:0
    Read Errors:0
    Write Errors:0
    Unmap Errors:0
    Checksum Errors:0
    Error Log Entries:0
  Critical Warnings:
    Temperature: OK
    Available Spare: OK
    Device Reliability: OK
    Read Only: OK
    Volatile Memory Backup: OK
```

tanabarr commented:
After the I/O server processes are stopped by executing `dmg system stop --force`:
- `dmg storage scan [--verbose]` output remains the same.
- `dmg storage scan --nvme-health` output is similar (will have been updated if any stats have changed since the last call) but there will be no `Timestamp` field displayed in the health stats.
- `dmg storage scan --nvme-meta` will reflect the fact that DAOS server metadata is only held on the NVMe controllers during the time that the I/O server processes exist.
```
bash-4.2$ dmg storage scan --nvme-meta
------------
wolf-[71-72]
------------
PCI:0000:81:00.0 Model:INTEL SSDPED1K750GA  FW:E2010435 Socket:1 Capacity:750 GB
  No SMD devices found

PCI:0000:da:00.0 Model:INTEL SSDPED1K750GA  FW:E2010435 Socket:1 Capacity:750 GB
  No SMD devices found

```

mjmac commented on src/proto/ctl/storage_nvme.proto:
```diff
@@ -65,13 +58,30 @@ message NvmeController {
+	}
+
+	// SmdDevice represents a blobstore created on a NvmeController_Namespace.
+	// TODO: this should really be embedded in Namespace above.
```
Why not just do it now, then?

tanabarr replied:
more changes would be required to pass the namespace identifier back alongside the Smd info and I didn't want to add further complexity to this PR, that would really be a cosmetic change as we don't really use the namespace info other than calculating cumulative capacity

mjmac reviewed (commented):
Generally looks good. Are there documentation changes needed for this update? Is there a ticket for removing the older commands (and converting tests that use them)?

When I was testing this locally, I found a problem after starting the ioservers:
```
[mjmac@linuxvm daos]$ dmg storage scan --verbose
---------
localhost
---------
        No SCM modules found

NVMe PCI     Model                FW Revision Socket ID Capacity
--------     -----                ----------- --------- --------
0000:03:00.0 VMware Virtual NVMe  1.0         0         8.6 GB
0000:0b:00.0 VMware Virtual NVMe  1.0         0         8.6 GB

[mjmac@linuxvm daos]$ dmg storage scan --nvme-meta
Errors:
  Hosts     Error                                                                                                                      
  -----     -----                                                                                                                      
  localhost updating bdev health and smd info: generate alt key for controller from health stats from smd device 42dd895f-007b-4c1c-a333-3a6c053a7f5f: missing model identifier
```

Not sure what's going on there, but the model shows up in the verbose output so it's not blank...

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/13/execution/node/61/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/14/execution/node/61/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/15/execution/node/61/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/16/execution/node/61/log

tanabarr commented:
> Generally looks good. Are there documentation changes needed for this update? Is there a ticket for removing the older commands (and converting tests that use them)?
> 
> When I was testing this locally, I found a problem after starting the ioservers:
> 
> ```
> [mjmac@linuxvm daos]$ dmg storage scan --verbose
> ---------
> localhost
> ---------
>         No SCM modules found
> 
> NVMe PCI     Model                FW Revision Socket ID Capacity
> --------     -----                ----------- --------- --------
> 0000:03:00.0 VMware Virtual NVMe  1.0         0         8.6 GB
> 0000:0b:00.0 VMware Virtual NVMe  1.0         0         8.6 GB
> 
> [mjmac@linuxvm daos]$ dmg storage scan --nvme-meta
> Errors:
>   Hosts     Error                                                                                                                      
>   -----     -----                                                                                                                      
>   localhost updating bdev health and smd info: generate alt key for controller from health stats from smd device 42dd895f-007b-4c1c-a333-3a6c053a7f5f: missing model identifier
> ```
> 
> Not sure what's going on there, but the model shows up in the verbose output so it's not blank...
> 
> Update: After running it again it worked. I'll see if I can reproduce this. Maybe it's a cache thing?
> 
> Updated update: After running it again several times I noticed that the smd info is duplicated multiple times:
> 
> ```
> [mjmac@linuxvm daos]$ dmg storage scan --nvme-meta
> ---------
> localhost
> ---------
> PCI:0000:03:00.0 Model:VMware Virtual NVMe  FW:1.0 Socket:0 Capacity:8.6 GB
>   SMD Devices
>     UUID:42dd895f-007b-4c1c-a333-3a6c053a7f5f Targets:[0] Rank:0 State:NORMAL
>     UUID:42dd895f-007b-4c1c-a333-3a6c053a7f5f Targets:[0] Rank:0 State:NORMAL
>     UUID:42dd895f-007b-4c1c-a333-3a6c053a7f5f Targets:[0] Rank:0 State:NORMAL
>     UUID:42dd895f-007b-4c1c-a333-3a6c053a7f5f Targets:[0] Rank:0 State:NORMAL
>     UUID:42dd895f-007b-4c1c-a333-3a6c053a7f5f Targets:[0] Rank:0 State:NORMAL
> 
> PCI:0000:0b:00.0 Model:VMware Virtual NVMe  FW:1.0 Socket:0 Capacity:8.6 GB
>   SMD Devices
>     UUID:fc24421d-4663-45ed-bff9-1212f485ee97 Targets:[0] Rank:1 State:NORMAL
>     UUID:fc24421d-4663-45ed-bff9-1212f485ee97 Targets:[0] Rank:1 State:NORMAL
>     UUID:fc24421d-4663-45ed-bff9-1212f485ee97 Targets:[0] Rank:1 State:NORMAL
>     UUID:fc24421d-4663-45ed-bff9-1212f485ee97 Targets:[0] Rank:1 State:NORMAL
>     UUID:fc24421d-4663-45ed-bff9-1212f485ee97 Targets:[0] Rank:1 State:NORMAL
> ```



> Generally looks good. Are there documentation changes needed for this update? Is there a ticket for removing the older commands (and converting tests that use them)?
> 
> When I was testing this locally, I found a problem after starting the ioservers:
> 
> ```
> [mjmac@linuxvm daos]$ dmg storage scan --verbose
> ---------
> localhost
> ---------
>         No SCM modules found
> 
> NVMe PCI     Model                FW Revision Socket ID Capacity
> --------     -----                ----------- --------- --------
> 0000:03:00.0 VMware Virtual NVMe  1.0         0         8.6 GB
> 0000:0b:00.0 VMware Virtual NVMe  1.0         0         8.6 GB
> 
> [mjmac@linuxvm daos]$ dmg storage scan --nvme-meta
> Errors:
>   Hosts     Error                                                                                                                      
>   -----     -----                                                                                                                      
>   localhost updating bdev health and smd info: generate alt key for controller from health stats from smd device 42dd895f-007b-4c1c-a333-3a6c053a7f5f: missing model identifier
> ```
> 
> Not sure what's going on there, but the model shows up in the verbose output so it's not blank...
> 
> Update: After running it again it worked. I'll see if I can reproduce this. Maybe it's a cache thing?
> 
> Updated update: After running it again several times I noticed that the smd info is duplicated multiple times:
> 
> ```
> [mjmac@linuxvm daos]$ dmg storage scan --nvme-meta
> ---------
> localhost
> ---------
> PCI:0000:03:00.0 Model:VMware Virtual NVMe  FW:1.0 Socket:0 Capacity:8.6 GB
>   SMD Devices
>     UUID:42dd895f-007b-4c1c-a333-3a6c053a7f5f Targets:[0] Rank:0 State:NORMAL
>     UUID:42dd895f-007b-4c1c-a333-3a6c053a7f5f Targets:[0] Rank:0 State:NORMAL
>     UUID:42dd895f-007b-4c1c-a333-3a6c053a7f5f Targets:[0] Rank:0 State:NORMAL
>     UUID:42dd895f-007b-4c1c-a333-3a6c053a7f5f Targets:[0] Rank:0 State:NORMAL
>     UUID:42dd895f-007b-4c1c-a333-3a6c053a7f5f Targets:[0] Rank:0 State:NORMAL
> 
> PCI:0000:0b:00.0 Model:VMware Virtual NVMe  FW:1.0 Socket:0 Capacity:8.6 GB
>   SMD Devices
>     UUID:fc24421d-4663-45ed-bff9-1212f485ee97 Targets:[0] Rank:1 State:NORMAL
>     UUID:fc24421d-4663-45ed-bff9-1212f485ee97 Targets:[0] Rank:1 State:NORMAL
>     UUID:fc24421d-4663-45ed-bff9-1212f485ee97 Targets:[0] Rank:1 State:NORMAL
>     UUID:fc24421d-4663-45ed-bff9-1212f485ee97 Targets:[0] Rank:1 State:NORMAL
>     UUID:fc24421d-4663-45ed-bff9-1212f485ee97 Targets:[0] Rank:1 State:NORMAL
> ```

The missing model identifier failure occurs when bio_monitor polling of the health log page has not yet occurred and therefore no stats exist to be returned, I've added a debug in the c mgmt module collection routine and a fault indicating that stats are not ready yet.

The duplication of SMD devices is because the cache gets updated and addition was not checking for duplicates, that's fixed.

A more subtle issue was that when starting and stopping I/O server's depending on the order of command invocation, sometimes cache was being updated with a limited list of bdevs (when some not available due to having been claimed), then on a subsequent command invocation updating over drpc would fail because cache wouldn't contain that controller entry. Fixed this by replacing req.Rescan with req.NoCache so , instead of updating the cache on rescan , cache is bypassed when necessary to obtain fresh stats/smd data but static controller info in cache does not get erased.

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/17/execution/node/61/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/18/execution/node/61/log

daosbuild1 commented:
Test stage Unit Test completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/18/execution/node/503/log

ravalsam commented:
Tried running this test locally and it looks good. Tried before and after storage format.
Observed the same issue as @tanabarr mention, that it takes some time to update the result after server starts until that period it reports the error (for min or two)

daosbuild1 commented on src/control/cmd/dmg/pretty/storage_test.go:
```diff
@@ -172,6 +172,176 @@ PCI:%s Model:%s FW:%s Socket:%d Capacity:%s
+					&control.HostStorage{
+						NvmeDevices: storage.NvmeControllers{
+							controllerC,
+							controllerD,
```
```suggestion
							controlled,
```

ashleypittman replied:
If you add controlled to the list of ignored words here you won't get this error.  In fact the code is currently clean so you probably should, or somebody else will have to do it in future.

https://github.com/daos-stack/daos/blob/master/ci/patch_src_in_place#L4

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3423/21/execution/node/61/log

tanabarr commented:
> Tried running this test locally and it looks good. Tried before and after storage format.
> Observed the same issue as @tanabarr mention, that it takes some time to update the result after server starts until that period it reports the error (for min or two)

During the period when health stats are unavailable immediately after I/O process start, the behaviour now is that the following messages appear in the control log:
```
Sep 25 10:55:22 wolf-71.wolf.hpdd.intel.com daos_server[22038]: daos_io_server:0 DAOS I/O server (v1.1.0) process 23354 started on rank 0 with 16 target, 0 helper XS, firstcore 1, host wolf-71.wolf.hpdd.intel.com.
<snip >
Sep 25 10:55:25 wolf-71.wolf.hpdd.intel.com daos_server[22038]: daos_io_server:1 DAOS I/O server (v1.1.0) process 23355 started on rank 3 with 16 target, 0 helper XS, firstcore 25, host wolf-71.wolf.hpdd.intel.com.
<snip >
Sep 25 10:56:03 wolf-71.wolf.hpdd.intel.com daos_server[22038]: DEBUG 10:56:03.089621 ctl_storage_rpc.go:310: received StorageScan RPC
Sep 25 10:56:03 wolf-71.wolf.hpdd.intel.com daos_server[22038]: DEBUG 10:56:03.089661 ctl_storage_rpc.go:168: instance 0 storage scan: only show bdev devices in config [0000:81:00.0]
Sep 25 10:56:03 wolf-71.wolf.hpdd.intel.com daos_server[22038]: instance 0: health stats from smd uuid 6144eda4-0c3d-494a-9b04-2c10e5e5f65d: data not yet available
Sep 25 10:56:03 wolf-71.wolf.hpdd.intel.com daos_server[22038]: DEBUG 10:56:03.090227 ctl_storage_rpc.go:168: instance 1 storage scan: only show bdev devices in config [0000:da:00.0]
Sep 25 10:56:03 wolf-71.wolf.hpdd.intel.com daos_server[22038]: instance 1: health stats from smd uuid 3e8af622-8132-403c-a03a-6c9e4a4e753e: data not yet available
Sep 25 10:56:05 wolf-71.wolf.hpdd.intel.com daos_server[22038]: DEBUG 10:56:05.251254 ctl_storage_rpc.go:324: responding to StorageScan RPC
Sep 25 10:56:27 wolf-71.wolf.hpdd.intel.com daos_server[22038]: DEBUG 10:56:27.222918 ctl_storage_rpc.go:310: received StorageScan RPC
Sep 25 10:56:27 wolf-71.wolf.hpdd.intel.com daos_server[22038]: DEBUG 10:56:27.222963 ctl_storage_rpc.go:168: instance 0 storage scan: only show bdev devices in config [0000:81:00.0]
Sep 25 10:56:27 wolf-71.wolf.hpdd.intel.com daos_server[22038]: DEBUG 10:56:27.223294 instance_drpc.go:244: instance 0: health stats from smd uuid 6144eda4-0c3d-494a-9b04-2c10e5e5f65d with key INTEL SSDPED1K750GAPHKS75050060750BGN->0000:81:00.0
Sep 25 10:56:27 wolf-71.wolf.hpdd.intel.com daos_server[22038]: DEBUG 10:56:27.223364 ctl_storage_rpc.go:168: instance 1 storage scan: only show bdev devices in config [0000:da:00.0]
Sep 25 10:56:27 wolf-71.wolf.hpdd.intel.com daos_server[22038]: DEBUG 10:56:27.223519 instance_drpc.go:244: instance 1: health stats from smd uuid 3e8af622-8132-403c-a03a-6c9e4a4e753e with key INTEL SSDPED1K750GAPHKS7505005J750BGN->0000:da:00.0
Sep 25 10:56:29 wolf-71.wolf.hpdd.intel.com daos_server[22038]: DEBUG 10:56:29.410246 ctl_storage_rpc.go:324: responding to StorageScan RPC
```
From the timestamps we can see it is approximately a minute between I/O start-up and health stats successfully being retrieved.

During the period where stats are unavailable, `dmg storage scan --nvme-meta` will return:
```
-------
wolf-71
-------
PCI:0000:81:00.0 Model:INTEL SSDPED1K750GA  FW:E2010325 Socket:1 Capacity:750 GB
  No SMD devices found

PCI:0000:da:00.0 Model:INTEL SSDPED1K750GA  FW:E2010325 Socket:1 Capacity:750 GB
  No SMD devices found
```
and `dmg storage scan --nvme-health` will fall back to displaying cached health statistics retrieved in scan performed during the start-up of the control plane (`daos_server`).

jbrosenz commented on doc/admin/administration.md:
```diff
@@ -9,7 +9,9 @@ control plane and will be documented in a future revision.
-- Query Per-Server Metadata: `dmg storage query (list-devices|list-pools)`
+- Query Per-Server Metadata:
+  - `dmg storage query (list-devices|list-pools)`
+  - `dmg storage scan --nvme-meta` shows mapping of metadata to NVMe controllers
```
I thought this might be a good place to add the note about 'During the period where stats are unavailable, dmg storage scan --nvme-meta will return ...' and 'dmg storage scan --nvme-health will fall back to displaying cached health stats..'

tanabarr replied:
that period of unavailability is eliminated in follow-up https://github.com/daos-stack/daos/pull/3528

kjacque commented on doc/admin/administration.md:
```diff
@@ -48,7 +50,9 @@ boro-11
 
-- Query Storage Device Health Data: `dmg storage query (device-health|target-health)`
+- Query Storage Device Health Data:
+  - `dmg storage query (device-health|target-health)`
+  - `dmg storage scan --nvme-health` shows NVMe controller health stats
```
It may be helpful in both the metadata and health cases to explain the difference between the query command and the scan command, and the circumstances under which you'd use one rather than the other.

tanabarr replied:
I'm trying to get some consensus on that myself, when I do I will update the admin guide

kjacque commented on src/control/server/ctl_storage_rpc_test.go:
```diff
@@ -546,6 +547,73 @@ func TestServer_CtlSvc_StorageScan_PostIOStart(t *testing.T) {
 				Scm: &ScanScmResp{State: new(ResponseState)},
 			},
 		},
+		"scan twice bdev meta with multiple io servers up": {
```
Not totally following what "scan twice" is testing. I would guess the use of the cache, but did something change between the two scans that would make it obvious that we're using the cache, not making the second call? If there's anything you can do to clarify that, I think it would be helpful for future developers (and ourselves in the future).

tanabarr replied:
there was a bug that added duplicate SMD entries to the cache with multiple scans, this test checks cache contents is not duplicated over multiple scans, added comment in follow-up https://github.com/daos-stack/daos/pull/3528

kjacque reviewed (approved):
My comments are minor, and could be addressed in a follow-on.

tanabarr commented:
intermittent failures in CI
- DAOS-5610 (tests timed out when starting libdaos)
- DAOS-4665 (pool create failed -1020 DER_HG)
- DAOS-5502 (aggregation basic test failures)
